### DISCUSSION TITLE
API路径问题

### DISCUSSION THREAD

**Discussion by quixote-liu** [flipped-aurora/gin-vue-admin#1031] [Q&A] [answered]
我发现后端代码中注册的路由路径前缀并没有/api/，但是在web页面访问的路径是有/api/前缀的，我尝试在后端代码里面找到相应的逻辑但是没有发现。例如：注册路径为 [POST] /authority/getAuthorityList，客户端请求路径为： [POST] /api/authority/getAuthorityList

**Comment by pnck** [comment, marked as answer]
那个/api/是proxy出来的，通常为了防止同源问题会用 nginx（dev时用vite）把后端的地址代理到前端server的host下。

举个例子：

-  后端server: localhost:`8888`
- 前端server: localhost:`8080`

此时在浏览器访问时，浏览器请求是发给 localhost:`8080`的。如果不配置代理，那么发起ajax时，浏览器直接去请求  localhost:`8888` 就会收不到数据，因为跨域了。因此用vite或nginx把 localhost:`8888`代理到 localhost:`8080`下：

- 前端server: localhost:8080/
- 后端server:localhost:8080/api/

此时前后端在同一个域下就没有同源问题了。
参考：

https://github.com/flipped-aurora/gin-vue-admin/blob/fc306beece65a3a83becff0411d81f088b65a042/web/.env.production#L5
https://github.com/flipped-aurora/gin-vue-admin/blob/fc306beece65a3a83becff0411d81f088b65a042/web/.env.development#L4
注意这个rewrite：
https://github.com/flipped-aurora/gin-vue-admin/blob/fc306beece65a3a83becff0411d81f088b65a042/web/vite.config.js#L65

----

补充一下： 
1. 浏览器请求 localhost:8080，请求发给vite（或nginx）
2. 然后反代检查请求url是否是  /api/xxx 开头，如果是，把它重写为/xxx并代理请求 localhost:8888/xxx
3. 后端接受 /xxx 的请求，即gin里注册的路径
